### PR TITLE
Wip/hughsie/firmware base uri

### PR DESCRIPTION
**Pull request by hughsie**

### CODE REVIEW

File: libfwupd/fwupd-remote.c [hunk 1]
```diff
@@ -320,15 +332,19 @@ fwupd_remote_load_from_filename (FwupdRemote *self,
...
-	/* all LOCAL remotes have to include a valid File */
+	/* all LOCAL remotes have to include a valid MetadataURI */
 	if (self->kind == FWUPD_REMOTE_KIND_LOCAL) {
-		g_autofree gchar *filename_cache = NULL;
-		filename_cache = g_key_file_get_string (kf, group, "File", error);
-		if (filename_cache == NULL)
-			return FALSE;
+		const gchar *filename_cache = metadata_uri;
+		if (g_str_has_prefix (filename_cache, "file://"))
+			filename_cache += 7;
```
Comment (superm1) [inline libfwupd/fwupd-remote.c]: To my point above, if you just looked for the prefix of file:// you could handle this whole section without needing to look for the kind.

File: libfwupd/fwupd-self-test.c [hunk 2]
```diff
@@ -121,6 +121,39 @@ fwupd_remote_download_func (void)
...
+			 "https://s3.amazonaws.com/lvfsbucket/downloads/firmware.xml.gz");
+	g_assert_cmpstr (fwupd_remote_get_metadata_uri_sig (remote), ==,
+			 "https://s3.amazonaws.com/lvfsbucket/downloads/firmware.xml.gz.asc");
+
+	firmware_uri = fwupd_remote_build_uri (remote, "http://bbc.co.uk/firmware.cab", &error);
+	g_assert_no_error (error);
+	g_assert (firmware_uri != NULL);
+	firmware_uri_str = soup_uri_to_string (firmware_uri, FALSE);
+	g_assert_cmpstr (firmware_uri_str, ==,
+			 "https://my.fancy.cdn/firmware.cab");
```
Comment (superm1) [inline libfwupd/fwupd-self-test.c]: I think you should instead be checking this with `fwupd_remote_build_firmware_uri `

File: data/remotes.d/fwupd.conf [hunk 3]
```diff
@@ -4,4 +4,4 @@
 Enabled=true
 Type=local
 Keyring=none
-File=@datadir@/fwupd/remotes.d/fwupd/metadata.xml
+MetadataURI=file://@datadir@/fwupd/remotes.d/fwupd/metadata.xml
```
Comment (hughsie) [inline data/remotes.d/fwupd.conf]: Agree, but after trying to test all the ways people could get this wrong I didn't fancy supporting `File=http://` or `Type=download` with `MetadataURI=file://` -- saying "this is the one way to do it" makes the testing matrix much smaller.

File: data/tests/firmware-base-uri.conf [hunk 4]
```diff
@@ -0,0 +1,6 @@
+[fwupd Remote]
+Enabled=true
+Type=download
```
Comment (hughsie) [inline data/tests/firmware-base-uri.conf]: Agreed.

File: libfwupd/fwupd-remote.c [hunk 5]
```diff
@@ -135,13 +137,20 @@ fwupd_remote_set_url (FwupdRemote *self, const gchar *url)
...
-		self->uri_asc = fwupd_remote_build_uri (self, url_asc, NULL);
+		self->metadata_uri_sig = g_strconcat (metadata_uri, suffix, NULL);
+		self->uri_asc = fwupd_remote_build_uri (self, self->metadata_uri_sig, NULL);
 		basename_asc = g_path_get_basename (soup_uri_get_path (self->uri_asc));
 		self->filename_asc = g_strdup_printf ("%s-%s", self->id, basename_asc);
 	}
 }
 
+/* note, this has to be set after MetadataURI */
+static void
```
Comment (hughsie) [inline libfwupd/fwupd-remote.c]: I think it's only important if you're using the now-deprecated fwupd_remote_get_uri() functions.

File: libfwupd/fwupd-remote.c (hunk 1 above)
Comment (hughsie) [inline libfwupd/fwupd-remote.c]: Agree; have a look at the 2nd fixup commit and tell me if you think that's sane. Thanks.

File: libfwupd/fwupd-self-test.c (hunk 2 above)
Comment (hughsie) [inline libfwupd/fwupd-self-test.c]: Agree, fixed. I wasn't sure of a better name tbh.

**Comment by superm1** [comment]
Lgtm now

**Pull request closed without merging**